Add multiple peers in Peers addition dialog. Closes #1563, #2245, #3133, #1419, #3287, #1419

ngosang opened the pull request:
Closes #1563, #2245, #3133, #1419, #3287 and #1419.

Before:
![](http://i.imgur.com/CwCQVBl.png)
After: (Qt4)
![](http://i.imgur.com/qpHXl9M.png)
After: (Qt5)
![](http://i.imgur.com/c898tfZ.png)

Chocobo1 commented:
:thumbsup: very useful!
I didn't test it yet, just some suggestions I thought of to make it more perfect:
* The title should be capitalized: `Peers Addition` or maybe `Add Peers` (pick one).
* The window should be resizeable vertically and horizontally as there would be potential text clipping for high DPI users, and the text edit area should follow the resize.
* Just glimpse through the code, ignore me if wrong, it seems there would be problems when validating IPv6 address, e.g. `2001:db8::ff00:42:8329`, also see: [ipv6 and port](http://serverfault.com/questions/205793/how-can-one-distinguish-the-host-and-the-port-in-an-ipv6-url).
* [QT5 only, added for note] The `Format: IPv4...` label can be removed by giving an example using [setPlaceholderText](http://doc.qt.io/qt-5/qtextedit.html#placeholderText-prop), like: `setPlaceholderText( "1.2.3.4:20000\n[2000::45]:20000" );`

ngosang commented:
> The title should be capitalized: Peers Addition or maybe Add Peers (pick one).

Changed to 'Add Peers'

> The window should be resizeable vertically and horizontally as there would be potential text clipping for high DPI users, and the text edit area should follow the resize.

Already it's.

> Just glimpse through the code, ignore me if wrong, it seems there would be problems when validating IPv6 address, e.g. 2001:db8::ff00:42:8329, also see: ipv6 and port.

Now the IPv6 format is [IPv6]:port (like in uTorrent and other apps). I changed the Copy peer function to copy the IPv6 with this format. I have tested with this IPs and it's working:
```
[2001:0DB8:0000:0000:0000:0000:1428:57ab]:3267
[2001:0DB8:0000:0000:0000::1428:57ac]:3268
[2001:0DB8:0:0:0:0:1428:57ad]:3269
[2001:0DB8:0::0:1428:57ae]:3270
[2001:0DB8::1428:57af]:3271
180.230.64.18:36500
1.186.77.9:51719
```
> [QT5 only, added for note] The Format: IPv4... label can be removed by giving an example using setPlaceholderText, like: setPlaceholderText( "1.2.3.4:20000\n[2000::45]:20000" );

I like it but for now we have to be compatible with Qt4. When support is dropped you could made this change in Add magnet, Add torrent... and this one.

Chocobo1 commented:
Something missing from my last comment: 
* When an input line (IP:port) cannot be validated/read, output the problem string in the QMessageBox::warning instead of giving a general error message.

Should be a trival change. The rest is good for me!

ngosang commented:
I can compile in my machine but Travis fails. When you have time can you help me to fix that?

Chocobo1 commented:
Try put `#include <QMap>` in peersadditiondlg.h

ngosang commented:
> When an input line (IP:port) cannot be validated/read, output the problem string in the QMessageBox::warning instead of giving a general error message.

![](http://i.imgur.com/GJ9SUC7.png)

Thanks

Chocobo1 commented:
(Don't forget to capitalize the title)

ngosang commented:
> (Don't forget to capitalize the title)

I think it's not neccesary. There is other MsgBox with the title "No peer entered". No Peer Entered???

glassez commented:
>>[QT5 only, added for note] The Format: IPv4... label can be removed by giving an example using setPlaceholderText, like: setPlaceholderText( "1.2.3.4:20000\n[2000::45]:20000" );

>I like it but for now we have to be compatible with Qt4. When support is dropped you could made this change...

And why not make different versions for Qt4/Qt5 using conditional compilation directives.
If you do not use the advanced capabilities of Qt5, then we do not proceed to use it. There is always a lot of people who will say, "Why do we need Qt5, because everything works fine?".

ngosang commented:
I'm not opposing, just saying that in other windows the help is displayed in this way (in Qt4 and Qt5). See Add magnet, Add torrent... If I make this dialog looks different in Qt5 and the other windows have the old style will be weird.
Anyway if you want I can make the changes.

Chocobo1 commented:
It's not the last inconsistency (if you made the change) in this program.... just joking.
Thinking differently, this could be the first step example for later UI fixups.

glassez commented:
>Thinking differently, this could be the first example for following UI fixups.

Join this.
Besides, you could change other windows too (if you want, of course).

glassez commented:
>I'm not opposing, just saying that in other windows the help is displayed in this way (in Qt4 and Qt5)

By and large, we now have only "compatible" with Qt5. We almost never use its "charms", as compatibility has come recently (I mean in master). But every time we add to the code something new, it's an excuse to try a new approach this new thing and remake similar to it old ones.

ngosang commented:
I'm making the changes but I don't know if **[2000::45]:20000** is a good example. Can you give me other?
![](http://i.imgur.com/qfoQhGF.png)

glassez commented:
IP addresses around us.
For example, fdb1:9973:dbba:0:ca6c:87ff:fe3f:8133.

Chocobo1 commented:
Is this better? (text in setPlaceholderText)
```
Format: IPv4:Port / [IPv6]:Port
Example:
1.2.3.4:1234
[2001:0db8::1428:57af]:4321
```

Pardon me, if I contradict my previous post.

ngosang commented:
I'm going to use this:
```
Example:
198.51.100.1:20000
[2001:f0d0:1002:51::4]:20000
```

ngosang commented:
rebased

sledgehammer999 commented on src/gui/properties/peersadditiondlg.h:
```diff
@@ -32,29 +32,28 @@
 
 namespace BitTorrent
 {
     struct PeerAddress;
```
I suppose this forward declaration doesn't work and you needed to include peerinfo,h, correct? If yes, remove this then. Otherwise remove the include.

glassez replied:
Header must be moved to peersadditiondlg.cpp

glassez replied:
I mean ```#include "core/bittorrent/peerinfo.h"```.

sledgehammer999 replied:
Unless QList<> doesn't work with forward declaration.

ngosang replied:
@glassez I can't do that.
```
In file included from moc_peersadditiondlg.cpp:9:0:
gui/properties/peersadditiondlg.h:45:18: error: ‘BitTorrent’ no se declaró en este ámbito
     static QList<BitTorrent::PeerAddress> askForPeersEndpoints();
                  ^
gui/properties/peersadditiondlg.h:45:41: error: el argumento de plantilla 1 es inválido
     static QList<BitTorrent::PeerAddress> askForPeersEndpoints();
                                         ^
gui/properties/peersadditiondlg.h:46:11: error: ‘BitTorrent’ no se declaró en este ámbito
     QList<BitTorrent::PeerAddress> m_peersList;
           ^
gui/properties/peersadditiondlg.h:46:34: error: el argumento de plantilla 1 es inválido
     QList<BitTorrent::PeerAddress> m_peersList;
```

sledgehammer999 commented on src/gui/properties/peersadditiondlg.cpp:
```diff
@@ -0,0 +1,109 @@
+    : QDialog(parent)
+{
+    setupUi(this);
+    connect(buttonBox, SIGNAL(rejected()), this, SLOT(reject()));
```
You don't need to do this. It is called automatically.

ngosang replied:
It is called automatically only if it's in the .ui file.

sledgehammer999 replied:
You're correct. Then you don't need to change this.

ngosang replied:
@sledgehammer999 I'm finishing this. What I should do?

glassez replied:
>It is called automatically only if it's in the .ui file.

Typically, such connections are created in the ui. Why this is here?

sledgehammer999 replied:
I suspect he just copied the code from `peeraddition.cpp`

ngosang replied:
Because I copied it from the addTrackers window (with the old code not yours) and it was a mess... I'm doing in ui now.

sledgehammer999 commented on src/gui/properties/peersadditiondlg.cpp:
```diff
@@ -0,0 +1,109 @@
+    QString invalid_peer;
+    foreach (const QString &peer, peers) {
+        if (peer.lastIndexOf(":") == -1) {
+            qDebug("Unable to parse the provided peer: %s", qPrintable(peer));
```
logger message instead. NORMAL type?

sledgehammer999 replied:
Forget this. I saw that you use a QMessageBox to inform the user.

sledgehammer999 commented on src/gui/properties/peersadditiondlg.cpp:
```diff
@@ -0,0 +1,109 @@
+        if (peer.lastIndexOf(":") == -1) {
+            qDebug("Unable to parse the provided peer: %s", qPrintable(peer));
+            all_valid = false;
+	    invalid_peer = peer;
```
indentation

sledgehammer999 commented on src/gui/properties/peersadditiondlg.h:
```diff
@@ -32,29 +32,28 @@
-
-    static BitTorrent::PeerAddress askForPeerAddress();
+    static QList<BitTorrent::PeerAddress> askForPeersEndpoints();
+    QList<BitTorrent::PeerAddress> peers_list;
```
`m_peersList` or `m_peers`

sledgehammer999 commented on src/gui/properties/peersadditiondlg.cpp:
```diff
@@ -0,0 +1,109 @@
+        QString str_ip_port = peer;
+        str_ip_port.remove(QChar('[')).remove(QChar(']'));
+        QString str_ip = str_ip_port.mid(0, str_ip_port.lastIndexOf(":"));
+        QString str_port = str_ip_port.mid(str_ip_port.lastIndexOf(":")+1);
```
Don't you need to first extract the ip/port and then remove the brackets(`[]`) ? If the user hasn't specified a port for the IPv6 address you'll extract the last bit as a port...

sledgehammer999 commented on src/gui/properties/peersadditiondlg.cpp:
```diff
@@ -0,0 +1,109 @@
+        if (ip.isNull()) {
+            qDebug("Unable to parse the provided IP: %s", qPrintable(str_ip));
+            all_valid = false;
+	    invalid_peer = peer;
```
indentation

sledgehammer999 commented on src/gui/properties/peersadditiondlg.cpp:
```diff
@@ -0,0 +1,109 @@
+        if (!ok || port < 1 || port > 65535) {
+            qDebug("Unable to parse the provided port: %s", qPrintable(str_port));
+            all_valid = false;
+	    invalid_peer = peer;
```
indentation

sledgehammer999 commented on src/gui/properties/peersadditiondlg.cpp:
```diff
@@ -0,0 +1,109 @@
+            break;
+        }
+        qDebug("Valid peer: %s", qPrintable(peer));
+	BitTorrent::PeerAddress addr;
```
indentation for these 3 lines too.

sledgehammer999 commented on src/gui/properties/peersadditiondlg.ui:
```diff
@@ -0,0 +1,61 @@
+&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
+p, li { white-space: pre-wrap; }
+&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:'Sans Serif'; font-size:9pt; font-weight:400; font-style:normal;&quot;&gt;
+&lt;p style=&quot;-qt-paragraph-type:empty; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px; font-family:'Sans'; font-size:10pt;&quot;&gt;&lt;br /&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
```
Reset the translatable option in Qt Creator. It will delete the above lines.
Since the textedit is empty, the translators will not have anything to translate anyway.

sledgehammer999 commented on src/gui/properties/peersadditiondlg.cpp:
```diff
@@ -0,0 +1,109 @@
+
+#if (QT_VERSION >= QT_VERSION_CHECK(5, 0, 0))
+    label_format->hide();
+    peers_txt->setPlaceholderText("Example:\n198.51.100.1:20000\n[2001:f0d0:1002:51::4]:20000");
```
IMO, it should have the same text as the label.

ngosang replied:
I don't mind but I need a clear answer.

sledgehammer999 replied:
Sorry for being ambiguous. Yes, please make it have the same text/example as the label.

ngosang replied:
@sledgehammer999 I'm finishing this. What I should do?

sledgehammer999 replied:
Didn't I answer you? `Format: IPv4:port / [IPv6]:port`

sledgehammer999 commented:
Don't forget to delete the peeraddition.cpp file and also remove it from the relevant .pri.
Btw, this is eligible for v3_2_x backporting.

glassez commented on src/gui/properties/peersadditiondlg.cpp:
```diff
@@ -0,0 +1,109 @@
+    return dlg.peers_list;
+}
+
+void PeersAdditionDlg::validateInput()
```
This method has been implemented too complicated.

ngosang replied:
@glassez I can try with RegEx. Any suggestion?

glassez replied:
I didn't mean that!
At least, peer parsing should be done in a separate function.

sledgehammer999 commented:
@ngosang I have answered to your questions already. Please reload the pages from github so you can view my comments. Sometimes github is stupid.

ngosang commented:
I did all changes you suggested.
@glassez I tried to simplify things, take a look.
@sledgehammer999 I did the backport with the old code #3150 but I'm not going to do more changes into it. The code is a mess but it works well. We all are making efforts in backport the code but it's necessary? How long will take the 3.3.0 release?

sledgehammer999 commented:
>How long will take the 3.3.0 release?

It depends. There are stuff to be done. Like ensure a proper migration from the old BT_backup files to the new files... And I am sure glassez wants to put some more changes in before releasing.

sledgehammer999 commented on src/gui/properties/peersadditiondlg.cpp:
```diff
@@ -0,0 +1,111 @@
+    BitTorrent::PeerAddress addr;
+    QString strIP, strPort;
+
+    if (peer.indexOf("[") != -1 && peer.indexOf("]:") != -1) { // IPv6
```
The opening bracket must be in the beginning.

ngosang replied:
Not according to this, only in the else: https://github.com/qbittorrent/qBittorrent/blob/master/CODING_GUIDELINES.md#a-multiple-tests

sledgehammer999 replied:
I think you misunderstood me. An IPv6 ip with port will always begin with an opening bracket(`[`). So instead test if the first character is an opening bracket in the string.

ngosang replied:
Done. About the other comment?

sledgehammer999 commented on src/gui/properties/peersadditiondlg.cpp:
```diff
@@ -0,0 +1,111 @@
+    if (peer.indexOf("[") != -1 && peer.indexOf("]:") != -1) { // IPv6
+        strIP = peer.split("]:")[0].remove(QChar('['));
+        strPort = peer.split("]:")[1];
+    } else if (peer.indexOf(":") != -1) { // IPv4
```
What happens if user enters just and IPv6 ip without a port? eg `2001:0db8:85a3:0000:0000:8a2e:0370:7334`
I think it is better to check if there is a dot, which means IPv4. If not treat the IP as IPv6 (and try to detect brackets)

ngosang replied:
I made the port always compulsory (as it's in the format label) because there isn't a default port. Each client has his own default port and most of them chose the port randomly.

sledgehammer999 commented:
>We all are making efforts in backport the code but it's necessary?

If something is difficult to backport then it should be left in master only. Normally I'll request simple things to be backported.

ngosang commented:
> Normally I'll request simple things to be backported.

You can take a look at the old code, it's completely different. I let you 2 comments.

ngosang commented:
> I made the port always compulsory (as it's in the format label) because there isn't a default port. Each client has his own default port and most of them chose the port randomly.

![](http://i.imgur.com/c6uNoIg.png)

sledgehammer999 commented on src/gui/properties/peersadditiondlg.cpp:
```diff
@@ -0,0 +1,114 @@
+    QString strIP, strPort;
+
+    if (peer[0] == '[' && peer.indexOf("]:") != -1) { // IPv6
+        strIP = peer.split("]:")[0].remove(QChar('['));
```
Don't perform the split twice. Just save in a temp var.

sledgehammer999 commented on src/gui/properties/peersadditiondlg.cpp:
```diff
@@ -0,0 +1,114 @@
+        strPort = peer.split("]:")[1];
+    }
+    else if (peer.indexOf(":") != -1) { // IPv4
+        strIP = peer.split(":")[0];
```
same here

sledgehammer999 commented:
For me that's the last comment. I'll also wait for @glassez to say it is ok, since he participated in the review.
(making the port compulsory is correct).

glassez commented on src/gui/properties/peersadditiondlg.cpp:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    bool allValid = true;
+    QStringList peers = peers_txt->toPlainText().trimmed().split("\n");
```
Unneeded variable.

glassez commented on src/gui/properties/peersadditiondlg.cpp:
```diff
@@ -0,0 +1,110 @@
+        return;
+    }
+
+    bool allValid = true;
```
Get rid of that.

glassez commented:
>Like ensure a proper migration from the old BT_backup files to the new files..

I think this is the only blocking issue. There are already enough changes for a new release.
>And I am sure glassez wants to put some more changes in before releasing.

Yes, my mission is not finished yet. But now it is not large. I want to do this gradually so as not to affect the release cycle.
Only new GeoIP Manager is for 3.3.0 release. And current PR's (those that I will publish before the release).

ngosang commented:
This PR is ready.
I prefer not delay much the release of version 3.3.0. I think we are trying to be too ambitious including too much in each release but meanwhile there are many known bugs in the stable releases. The most important softwares release a version every X months regardless of whether it contains many changes because the important thing is the bugs are fixed in a bounded time.

sledgehammer999 commented:
@glassez I am ready to merge. OK?

glassez commented on src/gui/properties/peerlistwidget.cpp:
```diff
@@ -179,15 +182,17 @@ void PeerListWidget::showPeerListMenu(const QPoint&)
+            if (torrent->connectPeer(addr)) {
+                qDebug("Adding peer %s...", qPrintable(addr.ip.toString()));
+                Logger::instance()->addMessage(tr("Manually adding peer %1...").arg(addr.ip.toString()));
+            } else {
```
Fix braces in if-else block.

glassez commented on src/gui/properties/peerlistwidget.cpp:
```diff
@@ -179,15 +182,17 @@ void PeerListWidget::showPeerListMenu(const QPoint&)
-    else {
-      qDebug("No peer was added");
+    QList<BitTorrent::PeerAddress> peersList = PeersAdditionDlg::askForPeersEndpoints();
+    if (!peersList.isEmpty()) {
```
Explicit test is needed only if empty list is special case.

ngosang replied:
@glassez It's for display the MsgBox only if not empty.
```
QMessageBox::information(0, tr("Peer addition"), tr("The peers were added to this torrent."));
```

glassez replied:
Ok. But then there is one problem. It may happen that the list is not empty, but no peer was added.

ngosang replied:
@glassez True but I think It never goin to happen because the IP/port are valid at this point. I don't know what are the reasons to not be accepted, maybe in the ip filter list? I tried with a lot of weird things and never fails. Any suggestion?

glassez replied:
1. Leave it as is.
2. Don't show message box at all.
3. To track whether peers were really added.

glassez commented on src/gui/properties/peersadditiondlg.cpp:
```diff
@@ -0,0 +1,103 @@
+    if (!ok || port < 1 || port > 65535)
+        return addr;
+
+    addr.ip = ip.toString();
```
Why toString()???

ngosang commented:
if @glassez is not able to find more problems it's finished.

sledgehammer999 commented on src/gui/properties/peerlistwidget.cpp:
```diff
@@ -159,7 +159,10 @@ void PeerListWidget::showPeerListMenu(const QPoint&)
     QString myport = m_listModel->data(m_listModel->index(row, PeerListDelegate::PORT)).toString();
     selectedPeerIPs << myip;
-    selectedPeerIPPort << myip + ":" + myport;
+    if (myip.length() > 15) // IPv6
```
Where does `15` come from? According to [this](https://en.wikipedia.org/wiki/IPv6#Address_representation) IPv6 addresses can be shortened so eg `::1` is a valid IPv6 address.
I propose searching for the presence of dots instead, which will denote IPv4.

sledgehammer999 commented on src/gui/properties/peerlistwidget.cpp:
```diff
@@ -179,16 +182,20 @@ void PeerListWidget::showPeerListMenu(const QPoint&)
+            peerCount++;
+        }
+        else {
+            QMessageBox::critical(0, tr("Peer addition"), tr("The peer %1 could not be added to this torrent.").arg(addr.ip.toString()));
```
This could ugly if the user has pasted a bunch of bad IPs. I propose here you just leave a logger message of type warning. Also setting a flag eg `bool invalidPeers`. Then when outside the loop check the flag a post this MessageBox saying "Some peers couldn't be added. Check the Log for details."

ngosang commented:
Changed.

sledgehammer999 commented on src/gui/properties/peerlistwidget.cpp:
```diff
@@ -179,16 +182,22 @@ void PeerListWidget::showPeerListMenu(const QPoint&)
     }
+    if (peerCount < peersList.length())
+        QMessageBox::information(0, tr("Peer addition"), tr("Some peers could not be added. Check the Log for details."));
+    if (peerCount > 0)
```
this should be 'else if` to avoid both messages to appear for the same input.

ngosang commented:
Changed.

ngosang commented:
@sledgehammer999 I think this is ready (if you don't have more changes). About the v3.2.0 PR #3196 I prefer to not merge that, you can close if you want.

ngosang commented:
With the last commit Contol+C copies in format IP:Port instead of IP only.

@sledgehammer999 is there any reason to not merge?

sledgehammer999 commented:
@ngosang Rebase. Also edit your commits to include the bug numbers.
PS: Don't spam multiple "Closes" in them. Just use `Closes #<num>, #<num> and #<num>`. I know it will auto-close only the first one. I'll close the rest. It is ugly to use multiple "Closes".

ngosang commented:
@sledgehammer999 ready.

sledgehammer999 commented:
Thanks for this. If it is easily backported, please do so.
